### ISSUE TITLE
Add support for datastream resources

### ISSUE THREAD

**Issue by jtjanchor** [hashicorp/terraform-provider-google#10810]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

I would love to see support for the following Datastream resources:

* streams
* private connections
* connection profiles

Datastream is generally available as of November 30, 2021. It would be great to have the ability to manage datastream resources via Terraform. 

### New or Affected Resource(s)

The new resources might look something like this:

* google_datastream_stream 
* google_datastream_private_connection
* google_datastream_connection_profile

### Potential Terraform Configuration

This is not exhaustive, but resources might look something like this:

```tf
resource "datastream_stream" "my_stream" {
  name = "my-stream"
  display_name = "My Stream"
  source_connection_profile_name = datastream_connection_profile.my_profile.name
  
  destination_config {
    cloud_storage_source_config {...}
  }
  backfill {...}
  source_config {
    mysql_source_config {...}
  }
}

resource "datastream_connection_profile" "my-profile" {
  name = "..."
  display_name = "My Profile"

  mysql_profile {...}
  private_connectivity {
    private_connection_name = datastream_private_connection.my_private_connection
  }
}

resource "datastream_private_connection" "my_private_connection" {
  name = "..."
  display_name = "My Private Connection"
  
  vpc_peering_config {
    subnet = ""
    vpc = ""
  }
}
```

### References

This might be a helpful starting point: https://cloud.google.com/datastream/docs/use-the-cloud-sdk

**Comment by arunsingh2803** [comment]
I have written a python based automation for end to end deployment of DataStream. You can use this as IaC or integrate it with existing IaC for DataStream till terraform module for DataStream comes
https://github.com/GoogleCloudPlatform/professional-services/tree/main/tools/datastream-deployment-python-automation

**Comment by trentkgUjet** [comment]
I would love this feature!

**Comment by toninator** [comment]
> I would love this feature!

So would I!

**Comment by spideyz0r** [comment]
Waiting for it!!

**Comment by joshtree41** [comment]
any update?

**Comment by trentkgUjet** [comment]
I'm going to give this feature a try myself.

**Comment by jlowe64** [comment]
I have done a bit of work on a terraform module using @arunsingh2803 's script posted above. I have ripped some of his features out and added some features of my own. Would anyone be interested in me publishing this and allowing you all to modify it to be a fully featured module?

**Comment by melinath** [comment]
b/233804820

**Comment by trentkgUjet** [comment]
@melinath  Oh excellent! Do you have anything to show yet? I have only gotten as far as familiarizing myself with the work needed. I've setup the environment for `magic-modules` and can build and compile the go code. Happy to yield this to someone with more experience.

**Comment by melinath** [comment]
I'll reference this ticket in all PRs so they'll show up in the comment thread as I make them :-) The first one is above my previous comment.

**Comment by Rstar1998** [comment]
Any updates on when we will be able to create datastream resources using terraform ? Have to created everything manually 😞

**Comment by Rstar1998** [comment]
Thanks @melinath and team  [ Stephen Lewis (Burrows) ] for providing terraform support to GCP datastream connection profile. This has save a lot of my time rather than doing things manually

**Comment by smalot** [comment]
How can we use this new terraform resource ?
Are we obliged to wait for a new release ?

**Comment by Rstar1998** [comment]
@smalot https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/datastream_connection_profile

**Comment by jlowe64** [comment]
@melinath just checking in here. I see you've contributed the connection profiles and private connections, but I don't see in the documentation how to direct a connection profile to use the private connection that is created. Am I missing something?

**Comment by jlowe64** [comment]
@melinath sorry, I literally just saw your most recent pull request is exactly this issue. I guess you're waiting on something internally to be fixed in the mean time?

**Comment by melinath** [comment]
the work on this ticket is ongoing. No worries, didn't forget the stream resource :-)

**Comment by equals215** [comment]
@melinath beautiful! If you ever need help on producing some code please ping me, I'd be more than happy to contribute to the gcp provider

**Comment by laksh-bhat** [comment]
Hey @melinath we are so happy this is now in terraform and can't wait for the stream resource too :) 

We've been running into issues when destroying private connections `resource "google_datastream_private_connection" "datastream_conn" {}` with the error
```
│ Error: Error when reading or editing PrivateConnection: googleapi: Error 400: Resource '"projects/relyance-internal/locations/us-central1/privateConnections/datastream-conn"' has nested resources. If the API supports cascading delete, set 'force' to true to delete it and its nested resources.
```
this resource may need to support the equivalent of `--force` flag

**Comment by melinath** [comment]
@laksh-bhat could you please open a new `bug` issue & fill out the fields there? Thanks!

**Comment by juleslenotte** [comment]
Still highly interested in having the datastream resources managed in terraform over here.

**Comment by lihan** [comment]
+1

**Comment by melinath** [comment]
Some work remaining on this.

**Comment by GusPrice** [comment]
Super exciting to have this resource available! I saw in your PRs:
> In particular I'm excluding CMEK fields and oracle & postgres (because those require specific handwritten tests) and oracle & postgres support (because that requires separate, non-automated testing.)

Is there an intent to add postgres support as a follow up?

**Comment by fryck** [comment]
Perfect !

I have an error when I try to update an existing stream:

```
Error: Error waiting for Updating Stream: {"@type":"type.googleapis.com/google.rpc.ErrorInfo","domain":"datastream.googleapis.com","metadata":{"message":"Some validations failed to complete successfully, see the full list in the operation metadata.","originalMessage":"","time":"2023-02-01T14:09:20Z","uuid":"05945f93-a23a-11ed-a1af-b2d514f7a161"},"reason":"VALIDATION_FAILURE"}
```

Thanks !

**Comment by roaks3** [comment]
The last PR for this work was merged yesterday, which includes Postgres and Oracle sources: https://github.com/GoogleCloudPlatform/magic-modules/pull/7175.

Marking this issue as closed as support for these resources has been added, but for those experiencing an issue or potential bug, please open a new issue so that we can track it separately.
